Alignment and Justification

hojjatabdollahi opened the pull request:
This adds Text Alignment and Justification.

Done:
- [x] Left
- [x] Right
- [x] Center
- [x] Justify
- [x] Add it to the libcosmic example

notgull commented:
This would close #76

hojjatabdollahi commented:
What should we do with trailing whitespaces when doing a justify-fill? Libreoffice just pretends that there is no whitespaces, even if you have a hundred of them on that one line. 

For now, I'm going to remove the last whitespace so normal texts (with single spaces between words) look good.

alerque commented:
The simplest rubric to start with is to remove breaking whitespaces, leave non-breaking whitespaces. It is rare that a non-breaking space would end up on a trailing edge of a line because they typically go between things you don't want broken up, but I've seen them used in used in combination with other spaces specifically to output space that doesn't get dropped. There are several layers of complexity in typesetting beyond this, especially depending on whether you want to do good typographic justification, but this is a place to start. After you break your lines trim any breaking spaces.

dhardy commented:
```rust
pub enum Align {
    Left,
    Right,
    Center,
    Justified,
}
```
Default alignment for right-to-left text should be `Right`, unlike LTR text. I suggest adding `Default` or `Auto` as an option.

You probably aren't interested, but you can compare my hacks for BiDi+alignment+wrapping [here](https://github.com/kas-gui/kas-text/blob/master/src/display/wrap_lines.rs).

jackpot51 commented:
Very good work! I will review later today.

hojjatabdollahi commented:
@alerque Here is a picture of the trailing spaces that I'm talking about, I have added 10 spaces between the last words in the first line to cause the line to wrap:
![Screenshot from 2023-02-23 08-34-01](https://user-images.githubusercontent.com/2222199/220955140-10147078-b789-40ac-9ae5-0512b71d855f.png)
My question is, would trimming all these spaces (more than one space) give the wrong information to the user? As Libreoffice does. If I have 10 trailing spaces, but they are not shown and the text looks neat, is that bad?


And here is when it happens to single spaces between words, which I'm going to fix:
![Screenshot from 2023-02-23 08-24-12](https://user-images.githubusercontent.com/2222199/220952578-f76d3bca-6f88-4cb6-810b-559239fb6dec.png)

I already handle spaces that happen at the end of a line when doing soft wrapping for a paragraph (like line 3 of the picture). But currently I'm not handling the spaces that would be pushed to the end because of the justification.

hojjatabdollahi commented:
@dhardy Good idea, but I haven't seen "Default" or "Auto" anywhere. What I'm thinking is to make the alignment an option, if that has been set to something then apply that, otherwise go with the default.

I will definitely take a look at your code. Thank you.

alerque commented:
This is definitely a case of WYSIWYG editors having a bad case of [Text Editing Hates You TOo](https://lord.io/text-editing-hates-you-too/). There is no good way to show the input data and the expected output presentation at the same time. You must make a trade off benefiting one way or the other.

If you are having the user *enter* text and manipulate it then showing the presentation view where spaces are (correctly) swallowed is counter productive. But if you want to show the expected visual effect of justification, showing he spaces would be counter productive. At the toolkit level I think you would need more than one mode because the *intended use cases* will be different.

Either way visually eluding line-ending spaces (and stretching the mid-line ones) is one thing, but selecting the text and copying it should produce the underlying input.

hojjatabdollahi commented:
I removed the alignment for `Buffer`, now you can set the alignment for  `Bufferline` only. This will enable setting alignment per paragraph. 

I updated `editor-libcosmic` so now if you press the alignment buttons in the toolbar they only affect the current paragraph.

Things to improve:
- Currently selecting a whitespace that has been stretched because of Justify-Fill will cause a glitch that your selection would be expanded to the whole line until you get to the next glyph. This shouldn't be hard to find and fix. 
- Also I feel like selecting text when the text is Align-Centered should highlight differently, but I'm not sure how it should look like.
- What should we do with multiple trailing whitespaces? Should we add another option like "Justify-Fill-Pretty" which would get rid of all the trailing whitespaces? I think the usecase for such thing is in a more advanced editor and is not needed for cosmic-text.
- Finally, it would be nice to update `editor-libcosmic` to apply the align command to the current selection (if you're selecting multiple paragraphs) and not just the current one paragraph.

But I don't think these are blocking issues. I can work on them later and try to polish the selection highlighting.

dhardy commented:
> What I'm thinking is to make the alignment an option

Semantically equivalent... but then you need either `text.set_align(Some(alignment))` or `text.clear_alignment()`. Whatever. I'm not sure I'd expect a text editor to use this `Align` type directly anyway.

> At the toolkit level I think you would need more than one mode because the intended use cases will be different.

Text handling in a text editor vs WYSIWYG processor like Word vs compiled document like PDF is so different that I'm not sure I'd expect the same tool to attempt to support everything (supertext, clickable hyperlinks, embedded TeX, SVG with selectable text, paragraph breaks, indentation, ...). In theory it's possible, sure, but where should you aim?

Justification suffers from this issue, but in practice justified text is not used in a source code editor, so swallowing extra spaces here is probably fine. (In fact, most source code editors even hide extra spaces at the end of a line before wrapping.)

jackpot51 commented on src/buffer_line.rs:
```diff
@@ -87,7 +89,26 @@ impl BufferLine {
+    ///
+    /// Will reset shape and layout if it differs from current alignment.
+    /// Returns true if the line was reset
+    pub fn set_align(&mut self, align: Align) -> bool {
```
I would recommend that either set_align takes an Option<Align>, or that the Align enum has a Default variant.

hojjatabdollahi replied:
I'll change this function signature to accept an option. I intentionally didn't do it because I couldn't think of an equivalent of "Default" or "None" as a UI element.

jackpot51 commented:
This looks good to me now.
